docs/design_proposals/sync-improvements.md

```diff
@@ -0,0 +1,265 @@
+  ```yaml
+  - from: '*.js'
+    inferTo: true
+    flatten: true        # ERR
+  - from: 'src/**/*.js'
+    inferTo: true
+    strip: src/          # ERR
+  ```
+- It should be an error to specify `strip` with a prefix which which does not match the static prefix of the `from` field.
+  ```yaml
+  - from: 'src/**/*.js'
+    to: dest/
+    strip: src/sub        # ERR
+  - from: 'src/**/*.js'
+    to: dest/
+    strip: app/           # ERR
+  ```
+
+Open questions about the configuration:
+
+**\<Is the syntax clear enough?\>**
+Is the syntax clear?
+Is the configuration intuitive to use?
+Is there any ambiguity which needs attention?
+
+Resolution: __Not Yet Resolved__
+
+**\<Does it improve clarity to have the `inferTo` field?\>**
+A different approach could be to fall back to inference, if the `to` field is left blank.
+This would have the advantage that there needs to be no validation about having `to` and `inferTo` in the same rule, which should cause an error.
+
+Resolution: __Not Yet Resolved__
+   
+**\<Error for `strip` and `flatten`\>**
+Given the new sync config, users may try to combine various flags in unintended ways.
+For example, should `strip` and `flatten` in the same rule always be an error?
+My intuition would say, that `strip+flatten` is the same as `flatten`, but this should to be discussed.
+
+Resolution: __Not Yet Resolved__
+   
+   
+#### Migration plan
+The new configuration supports all existing use-cases and is fully backwards compatible.
+An automatic schema upgrade shall be implemented.
+
+### Open Issues/Question
+
+**\<Overlapping copies\>** 
+Overlapping copies can be caused by at least two ways:
+
+1. The `COPY/ADD` instructions in the Dockerfile may copy to the same destination. For example:
+   ```dockerfile
+   COPY foo /foo
+   COPY bar/ /   # bar contains a file called "foo"
+   ```
+2. Two sync rules may copy files to the same destination. For example:
+   ```yaml
+   sync:
+   - from: foo
+     to: /baz
+   - from: bar
+     to: /baz
+   ```
+   
+Should such cases be treated as errors, warnings, or not be detected at all?
+
+Resolution: __Not Yet Resolved__
+
+**\<Transitive copies\>** 
+Multi-stage Dockerfiles can lead to quite contrived scenarios. For example:
+```dockerfile
+FROM scratch
+ADD foo baz
+FROM scratch
+COPY --from=0 baz bar
+```
+Ideally, the sync destination should be inferred as `foo -> bar`.
+
+This is quite an edge case and can easily be circumvented.
+However, it is quite complex to implement.
+Should we bother about this special case?
+   
+Resolution: __Not Yet Resolved__
+
+**\<Other builders\>**
+We don't know how to obtain a mapping of local paths to remote paths for other builders (jib, bazel).
+This may even have to be implemented upstream.
+Until we can support those builders, we to handle the case when a user tries to use inference with those builders.
+
+- Option 1: bail out with an error
```

Comment:
Bail out with "not yet supported error"
It would be great if we can bail out at the start but checking if "SyncMap" is implemented and the Builder who don't implement it throw an error.
Something like that.

Reply:
If we want to throw an error, we must do it upfront during the Skaffold pipeline config validation. Otherwise we cannot differentiate between dependencies that are not copied into the container (e.g. Dockerfile) and builders that can not provide destinations paths.

But I also prefer Option 1, because it does not surprise users.

Reply:
This situation seem similar to a situation when the destination cannot be inferred by the builder.
```
FROM scratch
  ADD foo baz
  RUN mv baz bar
FROM scratch
  COPY --from=0 bar bar
```

Reply:
+1 on error - preferably at schema validation time to fail fast but could be just inference time too